src/System.Private.CoreLib/src/System/Threading/Timer.cs

```diff
@@ -674,9 +674,9 @@ public ValueTask CloseAsync()
             }
         }
 
-        void IThreadPoolWorkItem.Execute() => Fire();
+        void IThreadPoolWorkItem.Execute() => Fire(isThreadPool: true);
 
-        internal void Fire()
+        internal void Fire(bool isThreadPool = false)
```

Comment:
All calls to this should be from thread pool threads. Could add an assert to `FireNextTimers()` that it's a thread pool thread, and always use `RunFromThreadPool()` in `CallCallback()`.

Reply:
One `Fire` is not called from the `Dispatch` loop